### ISSUE TITLE
Regexp queries normalizes input when targeting keyword fields, changing the meaning of regexps

### ISSUE THREAD

**Issue by yrodiere** [elastic/elasticsearch#80189]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.15.0 (container image from DockerHub)

**Plugins installed**: []

**JVM version** (`java -version`): Whatever is included in the container image.

**OS version** (`uname -a` if on a Unix-like system):

```
Linux yrodiere.redhat 5.14.13-200.fc34.x86_64 #1 SMP Mon Oct 18 12:39:31 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Description of the problem including expected versus actual behavior**:

When executing a regexp query on an analyzed field, the input regexp is not analyzed. That's fine.

However, when executing a regexp query on an normalized (keyword) field, the input regexp **is** normalized, which can change the meaning of the regexp.

For instance, if the normalizer includes the `pattern_replace` token filter from [this example from the reference documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-pattern-replace-charfilter.html#_example_configuration_15), which replaces dashes within numbers with underscores, then a regexp such as `[0-9]` will be turned into `[0_9]`, which has a completely different meaning.

Related (though IMO not a good idea, at least for regexp queries): #43759

**Steps to reproduce**:

```sh
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/' -d'{"mappings": {"properties": {"mykeyword": { "type": "keyword", "normalizer": "dashes-to-underscores" } } }, "settings": { "analysis": {"char_filter": { "dashes-to-underscores": {"type": "pattern_replace", "pattern": "(\\d+)-(?=\\d)", "replacement": "$1_" } }, "normalizer": { "dashes-to-underscores": { "type": "custom", "char_filter": ["dashes-to-underscores"] } } } } }'
echo
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/_doc/111?refresh=true' -d '{"mykeyword" : "111-111-111" }'
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/_doc/222?refresh=true' -d '{"mykeyword" : "222-222-222" }'
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/_doc/999?refresh=true' -d '{"mykeyword" : "999-999-999" }'
echo
curl -XPOST -H "Content-Type: application/json" 'localhost:9200/test/_search?pretty' -d '{"query": { "regexp": { "mykeyword": { "value": "[0-9_]+" } } } }'
```

Result from the query execution:

```
{
  "took" : 4,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "999",
        "_score" : 1.0,
        "_source" : {
          "mykeyword" : "999-999-999"
        }
      }
    ]
  }
}
```

Only doc `999` matched, while one would expect `111` and `222` to match as well.

**Provide logs (if relevant)**:

N/A

**Comment by markharwood** [comment]
The same is true of all term-level queries of which regexp is an example.
This is [a documented behaviour](https://www.elastic.co/guide/en/elasticsearch/reference/current/term-level-queries.html)

>Term-level queries still normalize search terms for keyword fields with the normalizer property. For more details, see normalizer.

I agree it seems inconsistent with all other field types but it is the way things work. 

> a regexp such as [0-9] will be turned into [0_9], which has a completely different meaning.

Now that is a concern. Arguably any normalisation should apply to regex content characters only and not any regex operators. Ensuring that is likely to be hard to implement though.

**Comment by yrodiere** [comment]
> Arguably any normalisation should apply to regex content characters only and not any regex operators [...] Potentially tricky

A sure way to introduce a minefield of bugs, in my opinion :)

> The simple solution would be to say we don't try normalize regexps in future, but that would be a backwards compatibility break with existing systems.

My thoughts exactly. Breaking backwards is annoying, but IMO it's defendable if it's to get a more correct behavior. Especially in a major version such as 8.

I suppose one could do things "the right way", like the Elasticsearch team usually does, by starting with an opt-in behavior change with warnings, then switching to opt-out with warnings in a later version, then removing the old behavior entirely in a later version.

It's definitely a judgment call, though.

**Comment by markharwood** [comment]
@romseygeek had the suggestion of adding a `normalize` flag to the `regexp` query which defaults to true.
In scenarios such as yours it can be set to `false` to avoid any side-effects of normalizing the pattern string.

**Comment by yrodiere** [comment]
That would definitely work for me.

I still think `false` should (ultimately) be the default (principle of least surprise and all that), but I'm sure other users would disagree, so... :shrug:

**Comment by markharwood** [comment]
> I'm sure other users would disagree, 

Yeah. All those security systems running out there with buttloads of rules in the form of regex patterns spotting suspicious activity will be pretty upset.
